tephra-core/src/main/java/co/cask/tephra/TransactionManager.java

```diff
@@ -954,15 +963,83 @@ private boolean doInvalidate(long writePointer) {
     }
     return false;
   }
+  
+  public boolean truncateInvalidTx(Set<Long> invalidTxIds) {
+    // guard against changes to the transaction log while processing
+    txMetricsCollector.gauge("truncateInvalidTx", 1);
+    Stopwatch timer = new Stopwatch().start();
+    this.logReadLock.lock();
+    try {
+      boolean success;
+      synchronized (this) {
+        ensureAvailable();
+        success = doTruncateInvalidTx(invalidTxIds);
+      }
+      appendToLog(TransactionEdit.createTruncateInvalidTx(invalidTxIds));
+      txMetricsCollector.gauge("truncateInvalidTx.latency", (int) timer.elapsedMillis());
+      return success;
+    } finally {
+      this.logReadLock.unlock();
+    }
+  }
 
+  private boolean doTruncateInvalidTx(Set<Long> invalidTxIds) {
+    LOG.info("Removing tx ids {} from invalid list", invalidTxIds);
+    boolean success = invalid.removeAll(invalidTxIds);
+    if (success) {
+      invalidArray = invalid.toLongArray();
+    }
+    return success;
+  }
+
+  public boolean truncateInvalidTxBefore(long time) throws InvalidTruncateTimeException {
+    // guard against changes to the transaction log while processing
+    txMetricsCollector.gauge("truncateInvalidTxBefore", 1);
+    Stopwatch timer = new Stopwatch().start();
+    this.logReadLock.lock();
+    try {
+      boolean success;
+      synchronized (this) {
+        ensureAvailable();
+        success = doTruncateInvalidTxBefore(time);
+      }
+      appendToLog(TransactionEdit.createTruncateInvalidTxBefore(time));
+      txMetricsCollector.gauge("truncateInvalidTxBefore.latency", (int) timer.elapsedMillis());
+      return success;
+    } finally {
+      this.logReadLock.unlock();
+    }
+  }
+  
+  private boolean doTruncateInvalidTxBefore(long time) throws InvalidTruncateTimeException {
+    LOG.info("Removing tx ids before {} from invalid list", time);
+    long truncateWp = TimeUnit.SECONDS.toMillis(time) * TxConstants.MAX_TX_PER_MS;
```

Comment:
Why did we decide on seconds as the unit for this argument, instead of milliseconds?  Sorry if we've already gone over this, just want to be clear.

Seems like milliseconds might actually be more intuitive in this case, since that is the standard way of referencing a timestamp in java.